src/tallies/tally_filter.cpp

```diff
@@ -0,0 +1,561 @@
+#include "openmc/tallies/tally_filter.h"
+
+#include <string>
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"  // for MAX_LINE_LEN;
+#include "openmc/error.h"
+#include "openmc/tallies/tally_filter_azimuthal.h"
+#include "openmc/tallies/tally_filter_cell.h"
+#include "openmc/tallies/tally_filter_cellborn.h"
+#include "openmc/tallies/tally_filter_cellfrom.h"
+#include "openmc/tallies/tally_filter_distribcell.h"
+#include "openmc/tallies/tally_filter_energyfunc.h"
+#include "openmc/tallies/tally_filter_legendre.h"
+#include "openmc/tallies/tally_filter_material.h"
+#include "openmc/tallies/tally_filter_mesh.h"
+#include "openmc/tallies/tally_filter_meshsurface.h"
+#include "openmc/tallies/tally_filter_mu.h"
+#include "openmc/tallies/tally_filter_polar.h"
+#include "openmc/tallies/tally_filter_sph_harm.h"
+#include "openmc/tallies/tally_filter_sptl_legendre.h"
+#include "openmc/tallies/tally_filter_surface.h"
+#include "openmc/tallies/tally_filter_universe.h"
+#include "openmc/tallies/tally_filter_zernike.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+std::vector<TallyFilterMatch> filter_matches;
+std::vector<TallyFilter*> tally_filters;
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+void
+free_memory_tally_c()
+{
+  #pragma omp parallel
+  {
+    filter_matches.clear();
+  }
+
+  for (TallyFilter* filt : tally_filters) {delete filt;}
+  tally_filters.clear();
+}
+
+//==============================================================================
+// C-API functions
+//==============================================================================
+
+// Fortran functions that will be called from C++
+extern "C" int verify_filter(int32_t index);
+extern "C" TallyFilter* filter_from_f(int32_t index);
+extern "C" void filter_update_n_bins(int32_t index);
+
+extern "C" {
+  int
+  openmc_cell_filter_get_bins(int32_t index, int32_t** cells, int32_t* n)
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "cell") {
+      set_errmsg("Tried to get cells from a non-cell filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto cell_filt = static_cast<CellFilter*>(filt);
```

Comment:
All of these static_casts should be dynamic_casts. In fact, you can use the resulting pointer as a way to check that the type is correct (rather than having to check `type()`) because the pointer will be null if it's not.